### PR TITLE
[flink] Add supports of lookup join on append-only table

### LINKED ISSUE #1410: [Feature] Append-only supports lookup join
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Append-only also can support Flink lookup join

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Aitozi**
### Purpose

Linked issue: close #1410

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+
+/** A {@link LookupTable} for table without primary key. */
+public class NoPrimaryKeyLookupTable implements LookupTable {
+
+    private final RocksDBSetState state;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java:38]: Maybe we should introduce `RocksDBMapState`?  Duplicate Rows should be output multiple times.

You should add test for this.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    @Override
+    public void refresh(Iterator<InternalRow> incremental) throws IOException {
+        while (incremental.hasNext()) {
+            InternalRow row = incremental.next();
+            joinKeyRow.replaceRow(row);
+            if (row.getRowKind() == RowKind.INSERT || row.getRowKind() == RowKind.UPDATE_AFTER) {
+                if (recordFilter.test(row)) {
+                    state.add(joinKeyRow, row);
+                } else {
+                    state.retract(joinKeyRow, row);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java:77]: Why we need to retract?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java (hunk 1 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java:38]: You are right, will fix this.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java (hunk 2 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java:77]: No need, will remove this.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java (hunk 1 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java]: I use the RocksDBListState here.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    public void refresh(Iterator<InternalRow> incremental) throws IOException {
+        while (incremental.hasNext()) {
+            InternalRow row = incremental.next();
+            joinKeyRow.replaceRow(row);
+            if (row.getRowKind() == RowKind.INSERT || row.getRowKind() == RowKind.UPDATE_AFTER) {
+                if (recordFilter.test(row)) {
+                    state.add(joinKeyRow, row);
+                }
+            } else {
+                state.retract(joinKeyRow, row);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java:80]: here we can just throw exception? We can assume all records should be insert or ua.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/RocksDBSetState.java [hunk 4]
```diff
@@ -89,7 +96,8 @@ public void retract(InternalRow key, InternalRow value) throws IOException {
     public void add(InternalRow key, InternalRow value) throws IOException {
         try {
             byte[] bytes = invalidKeyAndGetKVBytes(key, value);
-            db.put(columnFamily, writeOptions, bytes, EMPTY);
+            long count = toLong(db.get(columnFamily, bytes));
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/RocksDBSetState.java:99]: Why need to modify this class?

**Comment by JingsongLi** [comment]
Thanks @Aitozi for the update, left two comments.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java (hunk 3 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java:80]: as I  know, we can have delete record here, if we execute the `DELETE FROM` on a changlog table without primary key

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/RocksDBSetState.java (hunk 4 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/RocksDBSetState.java:99]: Reverted

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java (hunk 3 above)
Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java:80]: we can just not support `changlog table without pk`.

actually, I am considering to remove this mode....

Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/NoPrimaryKeyLookupTable.java]: Got it. Updated

**Review by JingsongLi — APPROVED** [review]
Looks good to me! Thanks @Aitozi

**Pull request merged**